src/java.base/share/classes/java/security/Signature.java

```diff
@@ -1005,17 +1005,18 @@ public final void setParameter(AlgorithmParameterSpec params)
     /**
      * Returns the parameters used with this signature object.
      *
-     * <p> If this signature has been initialized with parameters
-     * (by calling {@link #setParameter(AlgorithmParameterSpec)} or
-     * {@link #setParameter(String, Object)}) and the underlying signature
-     * implementation supports returning the parameters as
-     * {@code AlgorithmParameters}, this method returns the same parameters.
-     * If the parameters were not set, this method may return a combination
-     * of default and randomly generated parameter values if the
-     * underlying signature implementation supports it and can successfully
-     * generate them. Otherwise, {@code null} is returned.
+     * <p>The returned parameters may be the same that were used to initialize
+     * this signature, or may contain additional default or random parameter
+     * values used by the underlying signature implementation if the underlying
+     * signature implementation supports returning the parameters as
+     * {@code AlgorithmParameters}. If the required
+     * parameters were not supplied and the underlying signature implementation
+     * can generate the parameter values, it will be returned. Otherwise,
+     * {@code null} is returned.
```

Comment:
If we cannot guarantee anything but just want to clarify. how about just say "The returned parameters is a combination of user-provided values, default values, and/or random values used by the underlying signature implementation. It can also be {@code null} if there's none."

Reply:
I searched about "and/or" and it is said that "or" covers "and". So, "and/or" should just be "or".

I am on the fence for requiring provider to generate default parameters (using provider-specific or random values). Could there be scenarios where users are required to supply parameters? 
That aside, this javadoc is lastly updated by JDK-8243424. It seems that JDK's EdDSA signature impl has some interesting usage of AlgorithmParameterSpec which this javadoc has to cover/allow.

Reply:
RSASSA-PSS always requires a user-provided params.

I think one thing we can guarantee is that the default/random values generated by the impl will never overwrite the user-provided ones, they will only be supplemented. Also, I think the real usage of this method is that the result -- after converting to a spec -- can be set on the verify side and the verification should succeed.

I don't quite understand EdDSA. Looks like the params is not stored along with the signature in the algorithm identifier. I also haven't found different OIDs defined when prehash or context is used. How does the verifier know this kind of flavor settings?

Reply:
Right, the user-supplied values takes precedence and provider-specific default/random values should just be supplemental.

As for EdDSA, looks like the prehash and context are only in RFC 8032 and NOT RFC 8410.  caller has to pass these settings/values to the other entity through some other means since the getParameters() return null as there is no ASN.1 definition for the parameters. Looks like these values are packaged into a parameter spec and passed into the underlying EdDSA signature implementation in order to fit into existing API without adding new algorithm specific APIs.

Reply:
So, "the underlying signature implementation supports returning the parameters as {@code AlgorithmParameters}" is quite necessary. Xuelei's suggestion is OK, just change the last "and" to "or".

Reply:
I assume you were suggesting this? `"The returned parameters may be the same that were used to initialize this signature, or may contain additional default or random parameter values used by the underlying signature implementation. {https://github.com/code null} is returned if the required parameters were not supplied and the underlying signature implementation cannot generate the parameter values."`
But the "the underlying signature implementation supports returning the parameters as {https://github.com/code AlgorithmParameters}" is necessary. Strictly speaking, this is somewhat different than the "cannot generate parameter values" though. Perhaps we should go a bit broader for the last sentence regarding null return value?

Reply:
I suggest the last sentence to be "null is returned if the required parameters were not supplied **or** the underlying signature implementation cannot generate the parameter values." I used "or" because for EdDSA parameters are supplied but the impl cannot generate parameter values.

Reply:
The impl does not need to generate parameter values, but rather cannot convert the supplied parameter values into AlgorithmParameter objects.

Reply:
Then what does "cannot generate parameter values" mean? Any example?

Reply:
An example is RSASSA-PSS, i.e. it requires the caller to explicitly state which message digest to use, etc.

Reply:
You listed 2 cases when null is returned: 1) not supplied. 2) cannot generate. My understanding is that the RSASSA-PSS case is 1), and the EdDSA case is 2). This is why I suggested an "or" relation between them.

Reply:
For EdDSA, it's really the case of the signature impl cannot convert the supplied parameter values to "AlgorithmParameters" object.